### PR TITLE
Multi-dimensional flow model implementation based on SAM implementation

### LINKED ISSUE #23121: Incorporate SAM multi-D kernels into Navier-Stokes module
## Reason
We want to try to consolidate NEAMS thermal hydraulics development in common layers as upstream as possible, e.g. for example in the `navier_stokes` module

## Design
Move SAM multi-D kernels (and related objects such as bcs) into `navier_stokes`

## Impact
Create common layers for INL and ANL development. Let's all work in the same sandbox and create a united product as much as possible

### PR DESCRIPTION
**Pull request by lingzou**
This pull request includes multi-dimensional flow model implementation, both Navier-Stokes model and porous-medium model, based on existing SAM implementation

## Reason

The SAM multi-dimensional flow model is a generic CFD-like implementation which should be made available to public.

## Design

Based on existing SAM implementation.

## Impact

Users have more choices when selecting multi-dimensional flow model implementation.

Closes #23121

### CODE REVIEW

**Comment by lingzou** [comment]
@lindsayad There are a couple of minor things to be fixed but the pull request is ready for review.

**Comment by lindsayad** [comment]
@lingzou is it ok for me to push some commits to your branch? I see some style things I'd like to try to change like `map_var_number` to `mapVarNumber` just to be more consistent with naming in the repo

File: modules/navier_stokes/include/bcs/MDFluidEnergyBC.h
Comment (lindsayad) [inline modules/navier_stokes/include/bcs/MDFluidEnergyBC.h:15]:
```suggestion
/**
  * An integral BC for the energy (temperature) equation 
  */
```

File: modules/navier_stokes/include/bcs/MDFluidEnergyBC.h
Comment (lindsayad) [inline modules/navier_stokes/include/bcs/MDFluidEnergyBC.h:27]:
```suggestion
  virtual Real computeQpResidual() override;
  virtual Real computeQpJacobian() override;
  virtual Real computeQpOffDiagJacobian(unsigned int jvar) override;
```

File: modules/navier_stokes/include/bcs/MDFluidEnergyDirichletBC.h
Comment (lindsayad) [inline modules/navier_stokes/include/bcs/MDFluidEnergyDirichletBC.h:15]:
```suggestion
/**
  * A conditional Dirichlet BC for the energy (temperature) equation  
  */
```

File: modules/navier_stokes/include/bcs/MDFluidEnergyDirichletBC.h
Comment (lindsayad) [inline modules/navier_stokes/include/bcs/MDFluidEnergyDirichletBC.h:27]:
```suggestion
  virtual Real computeQpResidual() override;
  virtual Real computeQpJacobian() override;
  virtual Real computeQpOffDiagJacobian(unsigned int jvar) override;
```

File: modules/navier_stokes/include/bcs/MDFluidEnergyDirichletBC.h
Comment (lindsayad) [inline modules/navier_stokes/include/bcs/MDFluidEnergyDirichletBC.h:29]:
```suggestion
  bool shouldApply() override;
```

File: modules/navier_stokes/include/bcs/MDFluidIntegratedBCBase.h
Comment (lindsayad) [inline modules/navier_stokes/include/bcs/MDFluidIntegratedBCBase.h:56]:
```suggestion
  unsigned mapVarNumber(unsigned var);
```

File: modules/navier_stokes/include/bcs/MDFluidMassBC.h
Comment (lindsayad) [inline modules/navier_stokes/include/bcs/MDFluidMassBC.h:15]:
```suggestion
/**
  * A specific BC for the mass (pressure) equation 
  */
```

File: modules/navier_stokes/include/bcs/MDFluidMassBC.h
Comment (lindsayad) [inline modules/navier_stokes/include/bcs/MDFluidMassBC.h:27]:
```suggestion
  virtual Real computeQpResidual() override;
  virtual Real computeQpJacobian() override;
  virtual Real computeQpOffDiagJacobian(unsigned int jvar) override;
```

File: modules/navier_stokes/include/bcs/MDFluidMomentumBC.h
Comment (lindsayad) [inline modules/navier_stokes/include/bcs/MDFluidMomentumBC.h:25]:
```suggestion
  virtual Real computeQpResidual() override;
  virtual Real computeQpJacobian() override;
  virtual Real computeQpOffDiagJacobian(unsigned int jvar) override;
```

File: modules/navier_stokes/include/kernels/MDFluidEnergyKernel.h
Comment (lindsayad) [inline modules/navier_stokes/include/kernels/MDFluidEnergyKernel.h:18]:
```suggestion
```

File: modules/navier_stokes/include/kernels/MDFluidEnergyKernel.h
Comment (lindsayad) [inline modules/navier_stokes/include/kernels/MDFluidEnergyKernel.h:19]:
```suggestion
/**
  * The spatial part of the 3D energy conservation for fluid flow
  */
```

File: modules/navier_stokes/include/kernels/MDFluidKernelBase.h
Comment (lindsayad) [inline modules/navier_stokes/include/kernels/MDFluidKernelBase.h:34]: In general I try to encourage people (including myself who's often a guilty party) to move multi-line definitions out of the class definition, e.g. we could make this inline in the header outside of the class definition

File: modules/navier_stokes/include/kernels/MDFluidKernelBase.h
Comment (lindsayad) [inline modules/navier_stokes/include/kernels/MDFluidKernelBase.h:80]:
```suggestion
  /**
    * Helper function for mapping Moose variable numberings into
    * the "canonical" numbering for the porous medium equations.
    */
  unsigned int mapVarNumber(unsigned int var) const;
```

File: modules/navier_stokes/include/kernels/MDFluidKernelStabilization.h
Comment (lindsayad) [inline modules/navier_stokes/include/kernels/MDFluidKernelStabilization.h:14]:
```suggestion
/**
  * Base class for stabilization kernels.
  */
```

File: modules/navier_stokes/include/kernels/MDFluidMassKernel.h
Comment (lindsayad) [inline modules/navier_stokes/include/kernels/MDFluidMassKernel.h:14]:
```suggestion
/**
  * The spatial part of the 3D mass conservation for fluid flow
  */
```

File: modules/navier_stokes/include/kernels/MDFluidMomentumKernel.h
Comment (lindsayad) [inline modules/navier_stokes/include/kernels/MDFluidMomentumKernel.h:14]:
```suggestion
/**
  * The spatial part of the 3D momentum conservation for fluid flow
  */
```

File: modules/navier_stokes/include/kernels/PMFluidPressureTimeDerivative.h
Comment (lindsayad) [inline modules/navier_stokes/include/kernels/PMFluidPressureTimeDerivative.h:15]:
```suggestion
/**
  * The transient term of the porous-media mass conservation equation
  */
```

File: modules/navier_stokes/include/materials/MDFluidMaterial.h
Comment (lindsayad) [inline modules/navier_stokes/include/materials/MDFluidMaterial.h:16]:
```suggestion
/**
  * Fluid materials for 3D fluid model
  */
```

File: modules/navier_stokes/src/auxkernels/NSDensityAux.C
Comment (lindsayad) [inline modules/navier_stokes/src/auxkernels/NSDensityAux.C:10]:
```suggestion
#include "NSDensityAux.h"
#include "NS.h"
```

File: modules/navier_stokes/src/auxkernels/NSDensityAux.C
Comment (lindsayad) [inline modules/navier_stokes/src/auxkernels/NSDensityAux.C:21]:
```suggestion
  params.addRequiredCoupledVar(NS::pressure, "Coupled fluid pressure variable");
  params.addRequiredCoupledVar(NS::temperature, "Coupled fluid temperature variable");
```

**Review by lindsayad — DISMISSED** [review]
Some preliminary review through `include`. @lingzou I'd be happy to push commits to address these kind of cosmetic changes to help reduce time burden on you. Just let me know

**Comment by lingzou** [comment]
> 

@lindsayad Please go ahead do the necessary changes. Let me know if I have to do anything on my end, e.g., giving you access to the repo, things like that.

**Comment by lindsayad** [comment]
Sounds good!

**Comment by lindsayad** [comment]
@lingzou would it be possible to add a couple more tests? Currently it looks like the following objects aren't used:
- FluidWallMomentumBC
- MDFluidEnergyBC
- MDFluidEnergyDirichletBC
- MDFluidMomentumBC
- MDMomentumFreeSlipBC
I looked at the possibility of porting some of the existing tests in SAM that use these objects but it looks like they use the non-porous time derivative kernels so I would have to port those as well ... and it looks like some of those time derivative kernels use EOS APIs that don't have drop-in replacements with `SinglePhaseFluidProperties`, e.g. `enthalpy_from_T`. We have `h_from_p_T` but I didn't want to go too far down the rabbit hole of porting those kernels and changing to `SinglePhaseFluidProperties` and changing to `p_T` lookups ... as you may very well want to continue to use objects like `PBSodiumEquationOfState`

**Comment by lingzou** [comment]
> @lingzou would it be possible to add a couple more tests? Currently it looks like the following objects aren't used:
> 
> * FluidWallMomentumBC
> * MDFluidEnergyBC
> * MDFluidEnergyDirichletBC
> * MDFluidMomentumBC
> * MDMomentumFreeSlipBC
> 
> I looked at the possibility of porting some of the existing tests in SAM that use these objects but it looks like they use the non-porous time derivative kernels so I would have to port those as well ... and it looks like some of those time derivative kernels use EOS APIs that don't have drop-in replacements with `SinglePhaseFluidProperties`, e.g. `enthalpy_from_T`. We have `h_from_p_T` but I didn't want to go too far down the rabbit hole of porting those kernels and changing to `SinglePhaseFluidProperties` and changing to `p_T` lookups ... as you may very well want to continue to use objects like `PBSodiumEquationOfState`

@lindsayad thanks for keeping working on the PR. It is probably the best for me to add/port additional tests to cover these BCs. It shouldn't take long to do so, and we will let Elia understand that we are still working on it.

**Comment by lindsayad** [comment]
That would be great, thanks!

**Comment by lindsayad** [comment]
Ok @GiudGiud, @grmnptr I have commits on this PR so for SQA I cannot approve/merge it. Can one of you guys review this? Note that we cannot change the functionality of these objects as we need to maintain SAM compatibility and I'm not going to spend time on consolidation of the FE objects before this is merged per Elia's request

**Review by GiudGiud — COMMENTED** [review]
NSDensityAux should be deleted we have FluidDensityAux in the FP module
FluidWallMomentumBC is missing a prefix

Please link issues for:
- missing docs & docstrings
- issues attached to the original code
- missing Jacobian terms for that test that uses the FDP

Can we rename MD and PM? To INSFE and PINSFE most likely?

**Comment by lindsayad** [comment]
> Can we rename MD and PM? To INSFE and PINSFE most likely?

We could potentially do aliases. If we renamed we would break SAM

**Comment by GiudGiud** [comment]
Aliasing seems fine for a first step.

**Comment by lindsayad** [comment]
Ok aliasing doesn't seem like the correct solution. That only makes sense if you want the name in the `Registry` (and the name usable in the input file) to be different than the actual class name. It's not to support multiple allowable names in an input file. So we could deprecate the current names in favor of new names ... then both names would be usable in the input file but this would be denoting that we essentially want our users to adopt a new name eventually. @lingzou is this something that you think is acceptable?

**Comment by lingzou** [comment]
> ... that we essentially want our users to adopt a new name eventually. @lingzou is this something that you think is acceptable?

@lindsayad Yes, that will be fine, as long as we provide a time window for end users to adopt the new syntax/names. [I am at-ing @lingzou-anl so my ANL email receives update].

File: modules/navier_stokes/doc/content/source/materials/MDFluidMaterial.md
Comment (GiudGiud) [inline modules/navier_stokes/doc/content/source/materials/MDFluidMaterial.md:1]: rename?

File: modules/navier_stokes/include/bcs/INSFEFluidWallMomentumBC.h
Comment (GiudGiud) [inline modules/navier_stokes/include/bcs/INSFEFluidWallMomentumBC.h:14]: docstring

File: modules/navier_stokes/include/bcs/INSFEFluidMomentumBC.h
Comment (GiudGiud) [inline modules/navier_stokes/include/bcs/INSFEFluidMomentumBC.h:15]: same

File: modules/navier_stokes/include/kernels/PINSFEFluidTemperatureTimeDerivative.h
Comment (GiudGiud) [inline modules/navier_stokes/include/kernels/PINSFEFluidTemperatureTimeDerivative.h:15]: docstring

File: modules/navier_stokes/include/kernels/PINSFEFluidVelocityTimeDerivative.h
Comment (GiudGiud) [inline modules/navier_stokes/include/kernels/PINSFEFluidVelocityTimeDerivative.h:15]: same

File: modules/navier_stokes/include/materials/GenericPorousMediumMaterial.h
Comment (GiudGiud) [inline modules/navier_stokes/include/materials/GenericPorousMediumMaterial.h:14]: GenericPINSFEMaterial ?

File: modules/navier_stokes/doc/content/modules/navier_stokes/index.md
Comment (GiudGiud) [inline modules/navier_stokes/doc/content/modules/navier_stokes/index.md:31]: both have mixing length & seemingly others, see the MDFluidMaterial

File: modules/navier_stokes/test/tests/finite_element/pins/expansion-channel/tests
Comment (GiudGiud) [inline modules/navier_stokes/test/tests/finite_element/pins/expansion-channel/tests:9]:
```suggestion
    requirement = 'The system shall be able to solve a porous flow using slip-wall boundary condition.'
```

File: modules/navier_stokes/test/tests/finite_element/pins/channel-flow/tests
Comment (GiudGiud) [inline modules/navier_stokes/test/tests/finite_element/pins/channel-flow/tests:9]:
```suggestion
    requirement = 'The system shall be able to solve a porous flow with pressure drop due to viscous and inertia frictions.'
```

File: modules/navier_stokes/test/tests/finite_element/pins/channel-flow/tests
Comment (GiudGiud) [inline modules/navier_stokes/test/tests/finite_element/pins/channel-flow/tests:17]:
```suggestion
    requirement = 'The system shall be able to solve a porous flow with internal heating source.'
```

File: modules/navier_stokes/test/tests/finite_element/pins/channel-flow/tests
Comment (GiudGiud) [inline modules/navier_stokes/test/tests/finite_element/pins/channel-flow/tests:25]:
```suggestion
    requirement = 'The system shall be able to solve a porous flow when reverse flow happens.'
```

File: modules/navier_stokes/test/tests/finite_element/ins/lid_driven/tests
Comment (GiudGiud) [inline modules/navier_stokes/test/tests/finite_element/ins/lid_driven/tests:17]:
```suggestion
    requirement = 'The system shall be able to solve a canonical lid-driven problem using same order variables and the PSPG/SUPG stabilization scheme.'
```

**Review by GiudGiud — COMMENTED** [review]
ok let s get this in after these. 

In terms of coverage we could add two turbulence tests. 
I m ok with adding them later
Same for all object docs

**Review by GiudGiud — DISMISSED** [review]
5 last comments can be done in a follow up they are minor

**Comment by lindsayad** [comment]
Sorry @GiudGiud I pushed another commit with my airplane work, so will need another review

**Comment by lindsayad** [comment]
I tried to address the other 5 conversations

**Review by GiudGiud — APPROVED** [review]

**Pull request merged**
